libs/agno/agno/hooks/decorator.py

```diff
@@ -0,0 +1,164 @@
+from functools import wraps
+from typing import Any, Callable, TypeVar, Union, overload
+
+# Type variable for better type hints
+F = TypeVar("F", bound=Callable[..., Any])
+
+# Attribute name used to mark hooks for background execution
+HOOK_RUN_IN_BACKGROUND_ATTR = "_agno_run_in_background"
+
+
+def _is_async_function(func: Callable) -> bool:
+    """
+    Check if a function is async, even when wrapped by decorators like @staticmethod.
+    Traverses the full wrapper chain to find the original function.
+    """
+    from inspect import iscoroutinefunction, unwrap
+
+    # First, try the standard inspect function on the wrapper
+    if iscoroutinefunction(func):
+        return True
+
+    # Use unwrap to traverse the full __wrapped__ chain to the original function
+    try:
+        original_func = unwrap(func)
+        if original_func is not func and iscoroutinefunction(original_func):
+            return True
+    except ValueError:
+        # unwrap raises ValueError if it hits a cycle
+        pass
+
+    # Check if the function has CO_COROUTINE flag in its code object
+    try:
+        if hasattr(func, "__code__") and func.__code__.co_flags & 0x80:  # CO_COROUTINE flag
+            return True
+    except (AttributeError, TypeError):
+        pass
+
+    return False
+
+
+@overload
+def hook() -> Callable[[F], F]: ...
+
+
+@overload
+def hook(
+    *,
+    run_in_background: bool = False,
+) -> Callable[[F], F]: ...
+
+
+@overload
+def hook(func: F) -> F: ...
+
+
+def hook(*args, **kwargs) -> Union[F, Callable[[F], F]]:
+    """Decorator to configure hook behavior.
+
+    Args:
+        run_in_background: If True, this hook will be scheduled as a FastAPI background task
+                          when background_tasks is available, regardless of the agent/team's
+                          run_hooks_in_background setting. This allows per-hook control over
+                          background execution.
```

Comment:
Should we mention is intended as an AgentOS feature?

Reply:
Done